### PR TITLE
PRELIMINARY - Function call refactoring 

### PR DESCRIPTION
**Pull request by LucGenetier**
Separation of TexlFunction definitions & implementations
New IFunctionImplementation, InvokeAsync w/ IServiceProvider
New FunctionExecutionContext

To consider
- unify UDF
- consider IFunctionImplementation.InvokeAsync(FormulaValue[], IServiceProvider, CancellationToken)

### CODE REVIEW

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.ReflectionFunction.Invoke(System.IServiceProvider,Microsoft.PowerFx.Types.FormulaValue[])<br />❌CP0002 M:Microsoft.PowerFx.ReflectionFunction.InvokeAsync(System.IServiceProvider,Microsoft.PowerFx.Types.FormulaValue[],System.Threading.CancellationToken)

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs [hunk 1]
```diff
@@ -290,6 +290,8 @@ public SymbolTable GetMutableCopyOfFunctions()
 
         private protected readonly TexlFunctionSet _functions = new TexlFunctionSet();
 
+        internal virtual Dictionary<TexlFunction, IFunctionImplementation> FunctionImplementations { get; } = new ();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs:293]: >IFunctionImplementation [](http://example.com/codeflow?start=50&length=23)

A symbol table should be pure compile-time and not have runtime implementation on it.  So this mapping should probably be elsewhere.

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.PowerFxConfigExtensions.EnableMutationFunctions(Microsoft.PowerFx.SymbolTable)<br />❌CP0002 M:Microsoft.PowerFx.ReflectionFunction.Invoke(System.IServiceProvider,Microsoft.PowerFx.Types.FormulaValue[])<br />❌CP0002 M:Microsoft.PowerFx.ReflectionFunction.InvokeAsync(System.IServiceProvider,Microsoft.PowerFx.Types.FormulaValue[],System.Threading.CancellationToken)

**Comment by LucGenetier** [comment]
Closing due to lack of follow-up
Can't keep updating this PR

**Pull request closed without merging**